Add test command to tinygo

carolynvs opened the pull request:
Stuff that works so far:
* Added function to figure out the the package path based on the current directory so that we can call `tinygo test` without passing in a package path
* Test command passes flag to compile indicating that a test binary should be built
* Adds the test build tag
* Imports packages from test files

I'm stuck on how to find the TestMain that I have defined in my test file. It is coming back nil when I query for it.

You can try it out by running `make tinygo-test`.

After that, I'm not sure how I would reflect on the functions defined in the test files and then build up the test suite (M) in a basic block. First step I figured was calling a simple predefined test suite though. So that is what I'd like to understand before diving into other stuff.

This is for #228

deadprogram commented:
Hi @carolynvs thank you very much for working on this. One initial comment without looking at anything, please use the `dev` branch as your base since that is where all the active development is taking place. 

You may have already seen it, but take a look at https://github.com/tinygo-org/tinygo/blob/master/CONTRIBUTING.md#how-to-use-our-github-repository for more info.

I have taken the liberty of changing the branch for this PR to `dev` accordingly.

I will look at the content of this PR later today, thanks for getting us collectively started on this important set of features.

aykevl commented on src/testing/testing.go:
```diff
@@ -0,0 +1,54 @@
+"_syscall.Exit", referenced from:
+  _main in main.o
+ld: symbol(s) not found for architecture x86_64
+clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
`syscall.Exit` is implemented in the runtime, as can be seen here:

https://github.com/golang/go/blob/cb8aefd3b06f39679ebe1abbd38abef52cbd76b7/src/syscall/syscall.go#L102-L105

The implementation is here:

https://github.com/golang/go/blob/50bd1c4d4eb4fac8ddeb5f063c099daccfb71b26/src/runtime/runtime.go#L61

It could be implemented in TinyGo in the following way in the runtime package:

```go
//go:linkname exit exit
func exit(code int32)

//go:linkname syscall_Exit syscall.Exit
func syscall_Exit(cide int) {
    exit(int32(code))
}
```

What this basically does is implement `syscall.Exit` by calling the libc `exit` function.
You could make a separate PR for this feature.

carolynvs replied:
Thanks!

aykevl replied:
Reading it again, I see I made a mistake. The `exit` declaration should be the following:

```go
//go:export exit
func exit(code int32)
```

(`//go:export` is a bit of a misnomer, we're importing a function here. But it lets the compiler know it should follow the C ABI).

carolynvs replied:
I tried to get this to work, and while there aren't any errors, it doesn't emit the specified exit code either. Maybe someone else can pick this up and get it to work.

aykevl replied:
Hmmm maybe it should be `//go:extern exit`?

carolynvs replied:
Oh wait! I wasn't doing it right. I was testing it with `tinygo run` but after trying it with `tinygo build` then running the binary it created, that worked. So let me try more to get it to work with `tinygo test`.

aykevl replied:
So I thought it didn't work so I wrote my own: https://github.com/tinygo-org/tinygo/pull/380 (sorry, didn't see your comment, don't want to cause duplicate work...)
And had the exact same realization. We should probably fix this, it makes logical sense that `tinygo run` exits with the same exit code as the program you're trying to run. But that is somewhat of a separate issue.

aykevl commented:
Great to get this up! You haven't exactly chosen an easy feature to implement as a first thing.

I see a few changes in this PR that would be easier to review and merge as separate PRs, as they are not directly related to the `test` subcommand but are likely required for it:

* The `syscall.Exit` function I commented on.
* The change to the reflect package.
* Perhaps also the feature to leave out the package to build. For example, `tinygo flash -target=microbit` currently expects a package name. It would be useful to have this as a generic feature for `flash`, `run`, `build`, etc. and later `test`. It could be as simple as letting [this](https://github.com/tinygo-org/tinygo/blob/5569cd1b6b3aff8f52052d5ff1028d9bb87fba24/main.go#L522) default to `"."` if no package is given, instead of printing an error.

aykevl commented:
I've found another interesting file:
https://github.com/golang/go/blob/master/src/cmd/go/internal/test/test.go

Regarding the testing itself:
I think TestMain cannot be found because it is eliminated during the first round of dead code elimination. Anything not reachable from `main` is removed.
Also, I think testing should be implemented in an earlier stage, directly on the AST. As far as I know, the `go` toolchain does the same thing. This should also solve the DCE of TestMain. I think the logic should be something like this;

1. If there is no `main.TestMain`: generate it.
2. If there is a main function: delete it (for testing main packages).
3. Create a main function that calls `testing.MainStart` and then `main.TestMain` with the result.

This logic should likely live in the `loader` package. Take a look at the CGo support for how the AST can be modified.
After these modifications, the program can be compiled as usual.

carolynvs commented:
After rebasing on the `dev` branch (sorry about that!), I am seeing new behavior. When I run `tinygo test` I get the following output

```
error: import cycle:
	github.com/tinygo-org/tinygo/tests/example1
	testing
	runtime/trace
	context
	math/rand
	internal/testenv
	os/exec
	context
 at /usr/local/go/src/os/exec/exec.go:25:2
```

I am guessing that however I am building or running tinygo now that I'm on dev is no longer using the alternate source files in tinygo/src. So when it imports "testing", it is importing the real testing from go 1.11, which is then bringing in all those other packages. In tinygo/src/testing, it only imports fmt.

Do you have advice on what has changed, or what I was taking advantage of on master that no longer works on dev that is causing the wrong files to be used? 

Sorry I'm not quite sure if I should be saying that the GOROOT is wrong? It's hard coded to /usr/local/go (which is where go 1.11 is installed on my machine) but that was what master was doing too so I'm a bit unsure of how to describe the problem.

aykevl commented:
> After rebasing on the `dev` branch (sorry about that!)

Actually, I think that's a good thing. Especially right now when lots of stuff changes.

> When I run `tinygo test` I get the following output

Ah, yes, this was changed in https://github.com/tinygo-org/tinygo/pull/247. Standard library packages that must be overridden should now be explicitly listed here:

https://github.com/tinygo-org/tinygo/blob/a2d0f79be31b0122eeb237115aea4de6cb0c1c60/compiler/compiler.go#L199

carolynvs commented:
Thank you for the fast reply! That fixed it and I'm able to keep working on it. 👍

carolynvs commented:
FYI, the go toolchain used to do this at the SSA level it looks like. In 1.12 they've completely redone it, but here's the code for 1.11

https://github.com/golang/tools/blob/release-branch.go1.11/go/ssa/testmain.go

aykevl commented:
It's in the go/ssa package, but still they seem to produce regular Go code there:

https://github.com/golang/tools/blob/e9f45831faab4562ce3f53f7196b67c08f948e86/go/ssa/testmain.go#L210-L259

Very interesting how they make the test main package using a template.

deadprogram commented:
Now that v0.5.0 has been released, it would be great to get back to this PR. What do we need to do to support this?

carolynvs commented:
I actually got `tinygo test` working where the user supplied the TestMain (yeah I realize that's not the right signature, but it was the first step of what was needed)

```
func TestMain() {
	fmt.Println("running tests...")
	m := testing.M{
		Tests: []testing.TestToCall{
			{Name: "TestFoo", Func: TestFoo},
		},
	}
	m.Run()
}
```

Next step is to generate the TestMain if they don't have it (with the real signature), or call theirs, passing in the proper arguments if they do have one. Now that I have the right place in the loader, that is doable.

I'll get what I have working so far pushed up so you can see and keep working on it.

aykevl commented:
That's great! Maybe I can help with the TestMain generation?
Also related: compiling packages independently, see #285. That is a big change I want to do in smaller steps, and will definitely affect how `tinygo test` is implemented. For example, I want to make dead code elimination optional in preparation of splitting the compilation per package and caching the results.

carolynvs commented:
Oh shoot. At some point with the rebasing and updates/changes, I introduced a regression. When the code in the test files call a package, like `fmt.Println` I'm getting a null pointer exception. So I'm chasing down why  my test imports aren't working correctly anymore. Hopefully I can get that figured out tonight.

carolynvs commented:
Turns out my problem was just that I had rebased onto dev from a few weeks ago. I rebased to the most recent (post 0.5.0) and the regression went away. Phew! 😅

carolynvs commented:
Got a bit further tonight. I've updated the call to TestMain to supply a populated test suite, so it's finally calling the proper TestMain signature `TestMain(m *testing.M)`.

Unfortunately, I've hit a snag. The same code that worked in the TestMain, when moved into the main function and passed into TestMain as an argument is causing things to not compile properly.

```
$ tinygo test
DEBUG: TEST MAIN CONTENTS
package main

func main () {
	m := &testing.M{
		Tests: []testing.TestToCall{

			{Name: "TestFoo", Func: TestFoo},

		},
	}
	TestMain(m)
}

panic: interface conversion: ssa.Member is nil, not *ssa.Function

goroutine 1 [running]:
github.com/tinygo-org/tinygo/ir.(*Program).SimpleDCE(0xc0023a6c80)
	/Users/carolynvs/go/src/github.com/tinygo-org/tinygo/ir/passes.go:69 +0xb86
github.com/tinygo-org/tinygo/compiler.(*Compiler).Compile(0xc000001380, 0xc00001a1f8, 0x2b, 0xe, 0x0)
	/Users/carolynvs/go/src/github.com/tinygo-org/tinygo/compiler/compiler.go:255 +0x891
main.Compile(0xc00001a1f8, 0x2b, 0x4419c31, 0x4, 0xc00014a000, 0xc000eb3f08, 0x442f120, 0x0, 0x0)
	/Users/carolynvs/go/src/github.com/tinygo-org/tinygo/main.go:80 +0x47d
main.Test(0xc00001a1f8, 0x2b, 0x0, 0x0, 0xc0000c1f08, 0x0, 0xe)
	/Users/carolynvs/go/src/github.com/tinygo-org/tinygo/main.go:322 +0xe6
main.main()
	/Users/carolynvs/go/src/github.com/tinygo-org/tinygo/main.go:639 +0xbe2
```

I've pushed up all the latest code in case that helps. I am assuming that the error `interface conversion: ssa.Member is nil, not *ssa.Function` is due to `{Name: "TestFoo", Func: TestFoo},` in the new main that I'm generating, since `Func: TestFoo` is supposed to be a function pointer.

I tried adding a explicit import to the package being tested and changing that line to `Func: p.TestFoo` but that didn't help. 🤷‍♀️

aykevl commented on loader/loader.go:
```diff
@@ -182,12 +199,16 @@ func (p *Program) Parse() error {
 	}
 
+	if compileTestBinary {
+		p.SwapTestMain()
```
It looks like this returns an error that is not handled: `error: invalid source`.

carolynvs commented:
Ok, I hope enough of this is working now to merge and improve incrementally?

```
$ go install .
$ cd tests/example1
$ tinygo test
=== RUN   TestFail1
--- FAIL: TestFail1
	TestFail1 failed because of stuff and things

=== RUN   TestFail2
--- FAIL: TestFail2
	TestFail2 failed for reasons

=== RUN   TestPass
--- PASS: TestPass

exit status 2
FAIL
```

* Generates a TestMain and calls all Test* functions in the package.
* Followup: Improve signature matching to check that the Test* function has the correct function signature. I am just matching on the function name.
* Followup: Implement the gaps in tinygo so that we can use upstream testing package instead of the stub that I wrote.  I didn't spend time matching what's in testing because ultimately we should be working towards using this upstream package instead.
* Followup: Support the `tinygo test -run` flag to filter the tests to run.
* Followup: Implement syscall.Exit. I tried to follow your suggested implementation. It compiles and doesn't error out, but it also doesn't cause the supplied exit code to be returned either.

aykevl commented:
I have put these changes in the staging branch so that CircleCI will run all tests.

carolynvs commented:
I was able to get syscall.Exit to work, and have the exit code propagate from the test binary to `tinygo test` properly.  Just pushed that up too.

aykevl commented on src/testing/testing.go:
```diff
@@ -0,0 +1,77 @@
+}
+
+// Fatal is equivalent to Log followed by Fail
+func (t *T) Error(args ...interface{}) {
```
nit: the comment here has the wrong function name in it.

aykevl commented on target.go:
```diff
@@ -301,6 +301,30 @@ func getGopath() string {
 	return filepath.Join(home, "go")
 }
 
+func getPackageRoot() (string, error) {
```
In the meantime, I wrote my own not realizing that you already had one here:
https://github.com/tinygo-org/tinygo/blob/421ef04efb87889d73d5b404202a655f41d2e1e4/target.go#L319
I think it would be best to use that one instead.

aykevl replied:
Oh nevermind, `getPackageRoot` does something different.

carolynvs replied:
I will remove getPackageRoot and use "."

aykevl commented:
Can you run `make fmt`? That should format all the code using gofmt.

carolynvs commented:
Fixed!

deadprogram commented:
Is there anything stopping us from squashing/merging this PR now? Seems like there is already enough here to be useful.

deadprogram commented:
When I rebased this branch and `dev` I get the following errors:

```shell
$ make tinygo-test 
cd tests/tinygotest && tinygo test
# runtime
../../src/runtime/runtime_unix.go:22:6: exit redeclared in this block
../../src/runtime/runtime.go:111:6:     other declaration of exit
../../src/runtime/runtime_unix.go:83:6: syscall_Exit redeclared in this block
../../src/runtime/runtime.go:114:6:     other declaration of syscall_Exit
../../src/runtime/runtime_unix.go:84:7: cannot use code (variable of type int) as int32 value in argument to exit
Makefile:89: recipe for target 'tinygo-test' failed
make: *** [tinygo-test] Error 1
```

There are conflicting implementations of `exit` now, how should we proceed to resolve this?

deadprogram commented:
I commented the implementation of `exit` etc. in the `runtime.go` and was able to execute the tests within tinygo directory itself. Yes!

OK, then I went on to try to call some tests within the TinyGo drivers directory. I received the following error:

```shell
$ tinygo test ./thermistor
could not determine the package root from the current directory "/home/ron/Development/tinygo/drivers" and gopath "/home/ron/.gvm/pkgsets/go1.12/global"
```

The problem might be that I am using a symlink from my development directory to the actual gopath. I thought that we had that problem at one point, and it had had been resolved for `tinygo build`? If so that is the same thing we will need here probably.

aykevl commented:
@carolynvs can you rebase this branch on the current dev branch? There are a few conflicts.

deadprogram commented:
Oh, how much I want this feature!

deadprogram commented:
OK update to above comment. I need this feature now. How can I help?

aykevl commented:
Also, I should note that there is an upcoming big refactor that will conflict with this PR. It would be great to have this merged before then to avoid more (and harder to resolve) merge conflicts.

carolynvs commented on loader/loader.go:
```diff
@@ -62,6 +66,11 @@ func (p *Program) Import(path, srcDir string) (*Package, error) {
 	p.Packages[buildPkg.ImportPath] = pkg
+
+	if p.mainPkg == "" {
+		p.mainPkg = buildPkg.ImportPath
```
Is this a change that could be made in a follow-up PR? It's not actually based on the import statement order. The compiler calls ldprogram.Import or ldprogram.ImportFile, depending on whether it is testing a package or a single go file, that is how the mainPkg is being set.

I'm feeling pressured to get this merged quickly (understandable, sorry it's taken so long) and as you can tell I don't have time to work on this anymore. I tried to export and set the package from the compiler, but it doesn't have all the information to set it. So I'd have to replicate logic from the loader and this would slow down getting the PR merged even more.

carolynvs commented on main.go:
```diff
@@ -654,6 +680,14 @@ func main() {
 		err := Run(flag.Arg(0), *target, config)
 		handleCompilerError(err)
+	case "test":
+		pkgRoot, err := getPackageRoot()
```
I don't want to change the behavior of the other commands in this PR, it's out of scope. But I did change test to test if the positional argument is not present and use ".". I think that's what you were asking for, and would work for the other commands.

carolynvs commented on test.go:
```diff
@@ -0,0 +1,47 @@
+	"strings"
+)
+
+func RunTests() error {
```
I removed it since it wasn't used.

carolynvs commented:
I've rebased and addressed the review feedback with the exception of the following items:

* loader.mainPkg is still set as-is (see comment)
* build and other top level commands don't default to working against the current package when no positional argument is specified (out of scope for this PR)

deadprogram commented:
Hi @carolynvs agreed that this makes some very substantial progress. As you point out, still more is needed on this feature, but can be added in subsequent PRs. Most excellent contribution, thank you so much for putting in all the effort.

Now squashing/merging.
